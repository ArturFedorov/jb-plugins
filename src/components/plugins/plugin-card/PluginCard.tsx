import React, { FunctionComponent, useMemo } from 'react';
import { NavLink } from 'react-router-dom';
import { StarRating } from '../../common/star-rating/StarRating';
import { PluginIcon } from '../plugin-icon/PluginIcon';
import styles from '../plugins.module.scss';
import { IPlugin } from '../../../shared/interfaces/models/IPlugin';
import dateFormat from '../../../shared/utils/date-util/date.util';
import { formatNumber } from '../../../shared/utils/format-util/format.util';
import pluginIcon from '../../../assets/icons/jetbrains-logo.svg';

export interface IPluginCardProps {
  plugin: IPlugin;
}

export const PluginCard: FunctionComponent<IPluginCardProps> = ({ plugin }) => {
  const formattedDate = useMemo(() => dateFormat(plugin.date), [plugin.date]);
  const memoizedStarRating = useMemo(() => <StarRating rating={plugin.rating} />, [plugin.rating]);

  return (
    <NavLink to={`/plugins/${plugin.id}`} className={styles.pluginCardLink}>
      <div className={styles.pluginCard}>
        <div className={styles.pluginCardHeader}>
          <PluginIcon iconUrl={plugin.icon || pluginIcon} className={styles.pluginCardIcon} />
          <div className={styles.pluginCardSection}>
            <h3 className="is-lighter marginless">{plugin.name}</h3>
            {memoizedStarRating}
          </div>
        </div>
        <div className={styles.pluginCardDescription}>
          <span>{plugin.description}</span>
        </div>
        <div className={styles.pluginCardFooter}>
          <span className="is-caption">{formatNumber(plugin.downloads)} downloads</span>
          <span className="is-caption">{formattedDate}</span>
        </div>
      </div>
    </NavLink>
  );
};
