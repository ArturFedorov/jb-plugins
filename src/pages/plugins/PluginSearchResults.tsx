import React, { FunctionComponent } from 'react';
import { useLocation } from 'react-router-dom';

const PluginSearchResultsPage: FunctionComponent = () => {
  return <div>Search {useLocation().search}</div>;
};

export default PluginSearchResultsPage;
