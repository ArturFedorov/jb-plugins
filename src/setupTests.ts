// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import thunk from 'redux-thunk';
import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import createMockStore from 'redux-mock-store';

export const mockStore = createMockStore([thunk]);
configure({ adapter: new Adapter() });
