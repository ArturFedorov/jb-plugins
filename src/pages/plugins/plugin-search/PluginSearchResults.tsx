import React, { FunctionComponent, useEffect, useState } from 'react';
import classNames from 'classnames';
import { useHistory, useLocation } from 'react-router-dom';
import { connect } from 'react-redux';
import styles from './plugin-search.module.scss';
import { Input } from '../../../components/common/input/Input';
import { SearchIcon } from '../../../components/common/icons/SearchIcon';
import { fetchPlugins } from '../../../store/features/plugins/thunkActions';
import { RootState } from '../../../store/rootReducer';
import { getPlugins } from '../../../store/features/plugins/selectors';
import { getLoadingStatus } from '../../../store/features/base/selectors';
import { IPlugin } from '../../../shared/interfaces/models/IPlugin';
import { PluginList } from '../../../components/plugins/plugin-list/PluginList';
import { IPluginQueryParams } from '../../../shared/interfaces/api/IPluginQueryParams';
import { ArrowIcon } from '../../../components/common/icons/ArrowIcon';
import { Routes } from '../../../routes';

export interface ISearchResultsProps {
  plugins: IPlugin[];
  fetchPluginsConnect: (params?: IPluginQueryParams) => void;
  loadingFromApi: boolean;
}

const PluginSearchResultsPage: FunctionComponent<ISearchResultsProps> = ({
  plugins,
  loadingFromApi,
  fetchPluginsConnect
}) => {
  const DEFAULT_PAGE_SIZE = 12;
  const history = useHistory();
  const searchValue = new URLSearchParams(useLocation().search).get('query') || '';
  const [query, setQuery] = useState(searchValue);
  const [limit, setLimit] = useState(DEFAULT_PAGE_SIZE);

  useEffect(() => {
    fetchPluginsConnect({ query, limit });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fetchPluginsConnect, limit, searchValue]);

  const onChange = (val: string) => {
    setQuery(val);
  };

  const onEnter = () => {
    // rerun search set pagination to initial value
    setLimit(12);
    history.push(`${Routes.PLUGIN_SEARCH_RESULTS}?query=${query}`);
  };

  const showMore = () => {
    setLimit(limit + DEFAULT_PAGE_SIZE);
  };

  return (
    <div className={styles.pluginSearch}>
      <div className="container">
        <div className={styles.pluginSearchInput}>
          <h1>Search</h1>
          <Input
            value={query}
            placeholder="Search"
            icon={<SearchIcon />}
            onChange={onChange}
            onKeyEnter={onEnter}
          />
        </div>
        <div className={styles.pluginSearchList}>
          <h2 className={classNames(styles.pluginSearchHeader, 'is-lighter')}>
            Plugins {loadingFromApi}
          </h2>
          <PluginList isLoading={loadingFromApi} header="" plugins={plugins} isCentered={true} />
          {Boolean(plugins.length) && (
            <div className={styles.pluginSearchMore}>
              <span className={styles.pluginSearchButton} onClick={showMore}>
                <span className={styles.pluginSearchText}>Show More</span>
                <ArrowIcon viewBox="0 0 18 18" height={18} width={18} fill="#167dff" />
              </span>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

const mapDispatchToProps = {
  fetchPluginsConnect: fetchPlugins
};

const mapStateToProps = (state: RootState) => ({
  plugins: getPlugins(state.plugins),
  loadingFromApi: getLoadingStatus(state.base)
});

export default connect(mapStateToProps, mapDispatchToProps)(PluginSearchResultsPage);
