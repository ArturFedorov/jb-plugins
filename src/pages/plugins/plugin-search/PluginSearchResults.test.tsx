import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { mockStore } from '../../../setupTests';
import PluginSearchResults from './PluginSearchResults';
import { TestUtil } from '../../../shared/utils/test.util';
import { PluginCard } from '../../../components/plugins/plugin-card/PluginCard';

describe('PluginSearchResults', () => {
  const plugins = TestUtil.mockPlugins(4);
  const history = createMemoryHistory();
  const testInputValue = 'Component 1';

  const state = {
    plugins: {
      plugins
    },
    base: {
      loading: false
    }
  };

  const store = mockStore(() => state);

  const component = mount(
    <Provider store={store}>
      <Router history={history}>
        <PluginSearchResults />
      </Router>
    </Provider>
  );

  it('should match a snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should display all cards if query is empty', () => {
    const cards = component.find(PluginCard);
    expect(cards).toHaveLength(plugins.length);
  });

  it('should update query on enter', () => {
    const input = component.find('input');
    input.simulate('change', { target: { value: testInputValue } });
    input.simulate('keypress', { key: 'Enter', target: { value: testInputValue } });
    expect(history.location.search).toEqual(`?query=${testInputValue}&limit=12`);
  });
});
