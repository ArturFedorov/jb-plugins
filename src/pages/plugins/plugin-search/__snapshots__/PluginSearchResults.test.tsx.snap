// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginSearchResults should match a snapshot 1`] = `ReactWrapper {}`;
